File: src/agents/character.py
# -*- coding: utf-8 -*-
"""
Character Agent - 角色管理

负责角色档案的查询、关系分析和一致性检查。
从 Graph 存储中获取角色信息和关系网络。
"""

from typing import List, Dict, Any, Optional
from pydantic import BaseModel, Field
from .base import BaseAgent


class CharacterProfile(BaseModel):
    """角色档案"""
    name: str = Field(..., description="角色名称")
    role: str = Field(default="", description="角色定位: protagonist/antagonist/supporting")

    # 四个自我模型
    social_self: str = Field(default="", description="社会自我：外界看到的形象")
    personal_self: str = Field(default="", description="个人自我：自我认知")
    private_self: str = Field(default="", description="私密自我：隐藏的一面")
    hidden_self: str = Field(default="", description="隐藏自我：连自己都不知道的")

    # 核心属性
    desire: str = Field(default="", description="核心渴望")
    fear: str = Field(default="", description="最大恐惧")
    flaw: str = Field(default="", description="性格缺陷")
    strength: str = Field(default="", description="核心优势")

    # 外在特征
    appearance: str = Field(default="", description="外貌描述")
    speech_pattern: str = Field(default="", description="说话方式")
    mannerisms: List[str] = Field(default_factory=list, description="习惯动作")

    # 关系
    relationships: Dict[str, str] = Field(default_factory=dict, description="与其他角色的关系")


class CharacterContext(BaseModel):
    """角色上下文"""
    main_character: Optional[CharacterProfile] = None
    present_characters: List[CharacterProfile] = Field(default_factory=list)
    relationship_dynamics: List[str] = Field(default_factory=list, description="当前场景的关系动态")
    dialogue_guidelines: Dict[str, str] = Field(default_factory=dict, description="对话指南")


class CharacterAgent(BaseAgent):
    """
    角色 Agent

    职责：
    1. 查询角色档案
    2. 分析角色关系
    3. 验证角色行为一致性
    4. 提供对话风格指南
    """

    def __init__(self, llm=None, graph_engine=None, name: str = "Character", config: Optional[Dict[str, Any]] = None):
        super().__init__(name, config)
        self.llm = llm
        self._graph_engine = graph_engine

    @property
    def graph_engine(self):
        """延迟加载 Graph Engine"""
        if self._graph_engine is None:
            try:
                from src.graph.graph_engine import GraphEngine
                self._graph_engine = GraphEngine()
            except Exception as e:
                self.log_activity(f"Failed to load GraphEngine: {e}", level="WARNING")
        return self._graph_engine

    async def get_context(self, scene_info: Dict[str, Any]) -> CharacterContext:
        """
        获取场景所需的角色上下文

        Args:
            scene_info: 场景信息，包含 pov_character, characters 等

        Returns:
            CharacterContext: 角色上下文
        """
        pov_character = scene_info.get("pov_character", "")
        character_names = scene_info.get("characters", [])

        main_character = None
        present_characters = []
        relationship_dynamics = []
        dialogue_guidelines = {}

        # 获取 POV 角色档案
        if pov_character:
            main_character = await self._get_character_profile(pov_character)
            if main_character:
                dialogue_guidelines[pov_character] = self._generate_dialogue_guide(main_character)

        # 获取其他在场角色
        for name in character_names:
            if name != pov_character:
                profile = await self._get_character_profile(name)
                if profile:
                    present_characters.append(profile)
                    dialogue_guidelines[name] = self._generate_dialogue_guide(profile)

        # 分析关系动态
        if main_character and present_characters:
            relationship_dynamics = await self._analyze_relationships(
                main_character, present_characters
            )

        context = CharacterContext(
            main_character=main_character,
            present_characters=present_characters,
            relationship_dynamics=relationship_dynamics,
            dialogue_guidelines=dialogue_guidelines
        )

        self.log_activity(f"Generated character context: {pov_character}, {len(present_characters)} others")
        return context

    async def _get_character_profile(self, name: str) -> Optional[CharacterProfile]:
        """获取角色档案"""
        if not self.graph_engine:
            # 返回空档案
            return CharacterProfile(name=name)

        query = f"""
        MATCH (c:Character {{name: '{name}'}})
        OPTIONAL MATCH (c)-[r]->(other:Character)
        RETURN c, collect({{type: type(r), target: other.name}}) as relationships
        """

        try:
            result = self.graph_engine.query(query)
            if result and len(result) > 0:
                node = result[0]
                char_data = node.get("c", {})
                rels = node.get("relationships", [])

                relationships = {}
                for rel in rels:
                    if rel.get("target"):
                        relationships[rel["target"]] = rel.get("type", "KNOWS")

                return CharacterProfile(
                    name=name,
                    role=char_data.get("role", "supporting"),
                    social_self=char_data.get("social_self", ""),
                    personal_self=char_data.get("personal_self", ""),
                    private_self=char_data.get("private_self", ""),
                    hidden_self=char_data.get("hidden_self", ""),
                    desire=char_data.get("desire", ""),
                    fear=char_data.get("fear", ""),
                    flaw=char_data.get("flaw", ""),
                    strength=char_data.get("strength", ""),
                    appearance=char_data.get("appearance", ""),
                    speech_pattern=char_data.get("speech_pattern", ""),
                    mannerisms=char_data.get("mannerisms", []),
                    relationships=relationships
                )
        except Exception as e:
            self.log_activity(f"Character query failed: {e}", level="WARNING")

        return CharacterProfile(name=name)

    async def _analyze_relationships(
        self,
        main: CharacterProfile,
        others: List[CharacterProfile]
    ) -> List[str]:
        """分析角色间关系动态"""
        dynamics = []

        for other in others:
            rel_type = main.relationships.get(other.name, "NEUTRAL")

            # 基于关系类型生成动态描述
            if rel_type in ["ENEMY", "RIVAL", "ANTAGONIST"]:
                dynamics.append(f"{main.name} 与 {other.name} 存在对立关系，对话应充满张力")
            elif rel_type in ["FRIEND", "ALLY", "LOVER"]:
                dynamics.append(f"{main.name} 与 {other.name} 关系亲密，对话可以更轻松")
            elif rel_type in ["MENTOR", "STUDENT"]:
                dynamics.append(f"{main.name} 与 {other.name} 是师徒关系，存在指导与学习")
            else:
                dynamics.append(f"{main.name} 与 {other.name} 关系中性，保持礼貌距离")

        return dynamics

    def _generate_dialogue_guide(self, profile: CharacterProfile) -> str:
        """生成角色对话指南"""
        guides = []

        if profile.speech_pattern:
            guides.append(f"说话方式: {profile.speech_pattern}")

        if profile.mannerisms:
            guides.append(f"习惯动作: {', '.join(profile.mannerisms[:3])}")

        if profile.social_self:
            guides.append(f"表面形象: {profile.social_self}")

        if profile.private_self:
            guides.append(f"内心想法: {profile.private_self}")

        return "; ".join(guides) if guides else "无特殊指南"

    async def validate_behavior(
        self,
        character_name: str,
        action: str,
        context: CharacterContext
    ) -> Dict[str, Any]:
        """
        验证角色行为是否符合其设定

        Args:
            character_name: 角色名
            action: 行为描述
            context: 角色上下文

        Returns:
            验证结果
        """
        issues = []
        suggestions = []

        # 找到对应角色
        profile = None
        if context.main_character and context.main_character.name == character_name:
            profile = context.main_character
        else:
            for char in context.present_characters:
                if char.name == character_name:
                    profile = char
                    break

        if not profile:
            return {"is_valid": True, "issues": [], "suggestions": ["角色档案未找到，跳过验证"]}

        # 检查行为与性格是否冲突
        if profile.fear and profile.fear.lower() in action.lower():
            issues.append(f"{character_name} 的恐惧是 '{profile.fear}'，但行为中似乎克服了它，需要铺垫")

        if profile.flaw and "完美" in action:
            issues.append(f"{character_name} 有缺陷 '{profile.flaw}'，行为不应过于完美")

        is_valid = len(issues) == 0

        return {
            "is_valid": is_valid,
            "issues": issues,
            "suggestions": suggestions
        }

    def run(self, input_data: Any) -> Any:
        """同步运行接口"""
        import asyncio
        return asyncio.run(self.get_context(input_data))

File: src-ts/agents/character.ts
/**
 * character.ts - Character context agent
 *
 * Migrated from src/agents/character.py
 *
 * Responsible for character profile queries, relationship analysis,
 * and behavioral consistency checks.
 */

import { BaseAgent } from './base';
import type { IAgentGraphEngine } from './base';
import { LifecycleStage } from './lifecycle-hooks';
import type { DynamicCharacterState } from '../narrative/character-depth';
import { createEmptyDynamicState, mergeDynamicState } from '../narrative/character-depth';

const STM_CAPACITY = 20;

// ── Interfaces (Pydantic BaseModel -> TS interface) ────────

export interface CharacterProfile {
  name: string;
  role: string;

  // Four selves model
  socialSelf: string;
  personalSelf: string;
  privateSelf: string;
  hiddenSelf: string;

  // Core attributes
  desire: string;
  fear: string;
  flaw: string;
  strength: string;

  // External traits
  appearance: string;
  speechPattern: string;
  mannerisms: string[];

  // Relationships
  relationships: Record<string, string>;
}

export interface CharacterContext {
  mainCharacter: CharacterProfile | null;
  presentCharacters: CharacterProfile[];
  relationshipDynamics: string[];
  dialogueGuidelines: Record<string, string>;
}

// ── Agent ──────────────────────────────────────────────────

export class CharacterAgent extends BaseAgent {
  private _graphEngine: IAgentGraphEngine | null;
  private _shortTermMemory: string[] = [];
  private _longTermMemory: string[] = [];
  private _dynamicState: DynamicCharacterState = createEmptyDynamicState();

  constructor(
    deps?: {
      graphEngine?: IAgentGraphEngine;
      config?: Record<string, unknown>;
    },
  ) {
    super('Character', deps?.config);
    this._graphEngine = deps?.graphEngine ?? null;

    // Register lifecycle hooks for structured reasoning
    this.addHook(LifecycleStage.PERCEPTION, async (ctx) => {
      ctx.memoryContext = this.getMemoryContext();
      ctx.dynamicState = this._dynamicState;
      return ctx;
    });

    this.addHook(LifecycleStage.REFLECTION, async (ctx) => {
      if (ctx.lastAction) {
        this.pushMemory(ctx.lastAction as string);
      }
      if (ctx.newGoals || ctx.newStates) {
        this._dynamicState = mergeDynamicState(
          this._dynamicState,
          (ctx.newGoals as string[]) ?? [],
          (ctx.newStates as string[]) ?? [],
          (ctx.lastAction as string) ?? '',
        );
      }
      return ctx;
    });
  }

  // ── STM / LTM Memory Management ───────────────────────────

  pushMemory(event: string): void {
    this._shortTermMemory.push(event);
    if (this._shortTermMemory.length > STM_CAPACITY) {
      const oldest = this._shortTermMemory.shift();
      if (oldest) {
        const summary = oldest.length > 100
          ? oldest.slice(0, 97) + '...'
          : oldest;
        this._longTermMemory.push(`[LTM] ${summary}`);
      }
    }
  }

  getMemoryContext(): string {
    const stmPart = this._shortTermMemory.length > 0
      ? `Recent events:\n${this._shortTermMemory.map((e, i) => `${i + 1}. ${e}`).join('\n')}`
      : '';
    const ltmPart = this._longTermMemory.length > 0
      ? `\nLong-term memories:\n${this._longTermMemory.slice(-10).join('\n')}`
      : '';
    return stmPart + ltmPart;
  }

  get dynamicState(): DynamicCharacterState {
    return this._dynamicState;
  }

  // ── Lazy graph engine access ─────────────────────────────

  get graphEngine(): IAgentGraphEngine | null {
    return this._graphEngine;
  }

  set graphEngine(engine: IAgentGraphEngine | null) {
    this._graphEngine = engine;
  }

  // ── Public API ───────────────────────────────────────────

  /**
   * Build the character context needed for a given scene.
   */
  async getContext(sceneInfo: Record<string, unknown>): Promise<CharacterContext> {
    const povCharacter = (sceneInfo['pov_character'] as string) ?? '';
    const characterNames = (sceneInfo['characters'] as string[]) ?? [];

    let mainCharacter: CharacterProfile | null = null;
    const presentCharacters: CharacterProfile[] = [];
    const relationshipDynamics: string[] = [];
    const dialogueGuidelines: Record<string, string> = {};

    // Fetch POV character profile
    if (povCharacter) {
      mainCharacter = await this.getCharacterProfile(povCharacter);
      if (mainCharacter) {
        dialogueGuidelines[povCharacter] = this.generateDialogueGuide(mainCharacter);
      }
    }

    // Fetch other present characters
    for (const name of characterNames) {
      if (name !== povCharacter) {
        const profile = await this.getCharacterProfile(name);
        if (profile) {
          presentCharacters.push(profile);
          dialogueGuidelines[name] = this.generateDialogueGuide(profile);
        }
      }
    }

    // Analyze relationship dynamics
    if (mainCharacter && presentCharacters.length > 0) {
      const dynamics = await this.analyzeRelationships(mainCharacter, presentCharacters);
      relationshipDynamics.push(...dynamics);
    }

    const context: CharacterContext = {
      mainCharacter,
      presentCharacters,
      relationshipDynamics,
      dialogueGuidelines,
    };

    this.logActivity(
      `Generated character context: ${povCharacter}, ${presentCharacters.length} others`
    );
    return context;
  }

  /**
   * Validate whether a character action is consistent with their profile.
   */
  async validateBehavior(
    characterName: string,
    action: string,
    context: CharacterContext,
  ): Promise<{ isValid: boolean; issues: string[]; suggestions: string[] }> {
    const issues: string[] = [];
    const suggestions: string[] = [];

    // Find the matching profile
    let profile: CharacterProfile | null = null;
    if (context.mainCharacter && context.mainCharacter.name === characterName) {
      profile = context.mainCharacter;
    } else {
      profile = context.presentCharacters.find(c => c.name === characterName) ?? null;
    }

    if (!profile) {
      return {
        isValid: true,
        issues: [],
        suggestions: ['Character profile not found, skipping validation'],
      };
    }

    // Check for conflicts between action and personality
    if (profile.fear && action.toLowerCase().includes(profile.fear.toLowerCase())) {
      issues.push(
        `${characterName}'s fear is '${profile.fear}', but the action seems to overcome it -- needs buildup`
      );
    }

    if (profile.flaw && action.includes('\u5B8C\u7F8E')) {
      issues.push(
        `${characterName} has the flaw '${profile.flaw}', actions should not be overly perfect`
      );
    }

    return {
      isValid: issues.length === 0,
      issues,
      suggestions,
    };
  }

  /**
   * Synchronous entry point (mirrors Python `run`).
   */
  run(inputData: unknown): Promise<CharacterContext> {
    return this.getContext(inputData as Record<string, unknown>);
  }

  // ── Private helpers ──────────────────────────────────────

  private async getCharacterProfile(name: string): Promise<CharacterProfile | null> {
    if (!this._graphEngine?.query) {
      return { ...EMPTY_PROFILE, name };
    }

    const query = [
      `MATCH (c:Character {name: '${name}'})`,
      'OPTIONAL MATCH (c)-[r]->(other:Character)',
      'RETURN c, collect({type: type(r), target: other.name}) as relationships',
    ].join('\n');

    try {
      const result = await this._graphEngine.query(query);
      if (result && result.length > 0) {
        const node = result[0] as Record<string, unknown>;
        const charData = (node['c'] as Record<string, unknown>) ?? {};
        const rels = (node['relationships'] as Array<Record<string, unknown>>) ?? [];

        const relationships: Record<string, string> = {};
        for (const rel of rels) {
          const target = rel['target'] as string;
          if (target) {
            relationships[target] = (rel['type'] as string) ?? 'KNOWS';
          }
        }

        return {
          name,
          role: (charData['role'] as string) ?? 'supporting',
          socialSelf: (charData['social_self'] as string) ?? '',
          personalSelf: (charData['personal_self'] as string) ?? '',
          privateSelf: (charData['private_self'] as string) ?? '',
          hiddenSelf: (charData['hidden_self'] as string) ?? '',
          desire: (charData['desire'] as string) ?? '',
          fear: (charData['fear'] as string) ?? '',
          flaw: (charData['flaw'] as string) ?? '',
          strength: (charData['strength'] as string) ?? '',
          appearance: (charData['appearance'] as string) ?? '',
          speechPattern: (charData['speech_pattern'] as string) ?? '',
          mannerisms: (charData['mannerisms'] as string[]) ?? [],
          relationships,
        };
      }
    } catch (e) {
      this.logActivity(`Character query failed: ${e}`, 'WARNING');
    }

    return { ...EMPTY_PROFILE, name };
  }

  private async analyzeRelationships(
    main: CharacterProfile,
    others: CharacterProfile[],
  ): Promise<string[]> {
    const dynamics: string[] = [];

    for (const other of others) {
      const relType = main.relationships[other.name] ?? 'NEUTRAL';

      if (['ENEMY', 'RIVAL', 'ANTAGONIST'].includes(relType)) {
        dynamics.push(`${main.name} and ${other.name} have an antagonistic relationship; dialogue should be tense`);
      } else if (['FRIEND', 'ALLY', 'LOVER'].includes(relType)) {
        dynamics.push(`${main.name} and ${other.name} are close; dialogue can be more relaxed`);
      } else if (['MENTOR', 'STUDENT'].includes(relType)) {
        dynamics.push(`${main.name} and ${other.name} share a mentor-student bond with guidance and learning`);
      } else {
        dynamics.push(`${main.name} and ${other.name} maintain a neutral, polite distance`);
      }
    }

    return dynamics;
  }

  private generateDialogueGuide(profile: CharacterProfile): string {
    const guides: string[] = [];

    if (profile.speechPattern) {
      guides.push(`Speech pattern: ${profile.speechPattern}`);
    }
    if (profile.mannerisms.length > 0) {
      guides.push(`Mannerisms: ${profile.mannerisms.slice(0, 3).join(', ')}`);
    }
    if (profile.socialSelf) {
      guides.push(`Public persona: ${profile.socialSelf}`);
    }
    if (profile.privateSelf) {
      guides.push(`Inner thoughts: ${profile.privateSelf}`);
    }

    return guides.length > 0 ? guides.join('; ') : 'No specific guidelines';
  }
}

// ── Empty profile constant ─────────────────────────────────

const EMPTY_PROFILE: Omit<CharacterProfile, 'name'> = {
  role: '',
  socialSelf: '',
  personalSelf: '',
  privateSelf: '',
  hiddenSelf: '',
  desire: '',
  fear: '',
  flaw: '',
  strength: '',
  appearance: '',
  speechPattern: '',
  mannerisms: [],
  relationships: {},
};
